acvm/src/lib.rs

```diff
@@ -51,6 +51,13 @@ pub enum OpcodeResolutionError {
     IncorrectNumFunctionArguments(usize, BlackBoxFunc, usize),
 }
 
+#[derive(Debug, PartialEq)]
+pub enum GateResolution {
+    Solved,                     // The gate is resolved
+    Stalled(OpcodeNotSolvable), // The gate is not solvable
+    InProgress,                 // The gate is not solvable but could resolved some witness
```

Comment:
```suggestion
#[derive(Debug, PartialEq)]
enum AtomicOpcodeResolution {
    /// The opcode is resolved
    Solved,
    /// The opcode is not solvable
    Stalled(OpcodeNotSolvable), 
}

#[derive(Debug, PartialEq)]
pub enum OpcodeResolution {
    /// The opcode is resolved
    Solved,
    /// The opcode is not solvable
    Stalled(OpcodeNotSolvable),
    /// The opcode is not solvable but could resolve some witnesses
    InProgress,                 
}

// TODO implement OpcodeResolution::from<AtomicOpcodeResolution>
```

My thinking is that we'd have something along these lines. `ArithmeticSolver`/`DirectiveSolver` would return an `AtomicOpcodeResolution` and then we convert to an `OpcodeResolution` inside the global `solve` method.

Reply:
In any case, this should be renamed to `OpcodeResolution` and have the comments changed to doc comments.